tooling/templatize/pkg/pipeline/run.go

```diff
@@ -206,200 +228,68 @@ func RunPipeline(pipeline *types.Pipeline, ctx context.Context, options *Pipelin
-	})
-
-	// check for cycles
-	for _, node := range nodes {
-		seen := []*stepExecutionContext{
-			node.Step,
-		}
-		if err := traverse(node, nodes, seen); err != nil {
-			return nil, err
-		}
+	resourceGroup, exists := graphCtx.ResourceGroups[node.ResourceGroup]
+	if !exists {
+		return fmt.Errorf("could not find resource group %s", node.ResourceGroup)
 	}
 
-	return nodes, nil
-}
-
-func traverse(node *stepExecutionNode, all []*stepExecutionNode, seen []*stepExecutionContext) error {
-	for _, child := range node.Children {
-		for _, previous := range seen {
-			if previous == child {
-				var cycle []string
-				for _, i := range seen {
-					cycle = append(cycle, fmt.Sprintf("%s/%s", i.ResourceGroup.Name, i.Delegate.StepName()))
-				}
-				return fmt.Errorf("cycle detected, reached %s/%s via %s", child.ResourceGroup.Name, child.Delegate.StepName(), strings.Join(cycle, " -> "))
-			}
-		}
-		chain := seen[:]
-		chain = append(chain, child)
-		var childNode *stepExecutionNode
-		for _, candidate := range all {
-			if candidate.Step == child {
-				childNode = candidate
-			}
-		}
-		if childNode == nil {
-			return fmt.Errorf("could not find child node %s/%s - programmer error", child.ResourceGroup.Name, child.Delegate.StepName())
-		}
-		if err := traverse(childNode, all, chain); err != nil {
-			return err
-		}
+	step, exists := graphCtx.Steps[node.ServiceGroup][node.ResourceGroup][node.Step]
+	if !exists {
+		return fmt.Errorf("could not find step %s/%s/%s", node.ServiceGroup, node.ResourceGroup, node.Step)
 	}
-	return nil
-}
 
-func executeNode(executor Executor, node *stepExecutionNode, ctx context.Context, options *PipelineRunOptions, outputs Outputs, executed sets.Set[types.StepDependency], lock *sync.RWMutex) error {
-	logger, err := logr.FromContext(ctx)
+	subscriptionID, err := options.SubsciptionLookupFunc(ctx, resourceGroup.Subscription)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to lookup subscription ID for %q: %w", resourceGroup.Subscription, err)
 	}
-	logger = logger.WithValues("resourceGroup", node.Step.ResourceGroup.Name, "step", node.Step.Delegate.StepName())
-	thisNode := types.StepDependency{ResourceGroup: node.Step.ResourceGroup.Name, Step: node.Step.Delegate.StepName()}
-	lock.RLock()
-	alreadyDone := executed.Has(thisNode)
-	lock.RUnlock()
-	if alreadyDone {
-		logger.Info("Skipping execution, as it has already happened.")
-		// our graph may converge, where many children need one parent - no need to re-execute then
-		return nil
+	target := &executionTargetImpl{
+		subscriptionName: resourceGroup.Subscription,
+		subscriptionID:   subscriptionID,
+		region:           options.Region,
+		resourceGroup:    resourceGroup.ResourceGroup,
 	}
 
-	output, err := executor(node.Step.Delegate, ctx, node.Step.Target, options, outputs)
+	output, err := executor(step, ctx, target, options, state)
 	if err != nil {
 		return err
 	}
 
-	lock.Lock()
+	state.Lock()
```

Comment:
Use `defer unlock` here.  Looks risky for panic.

Reply:
very short  (7 lines) critical section with no early returns? seems ok